File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void func(int i , vector<int> &ds, vector<vector<int> > &ans,vector<int>& candidates, int target){

        if(i>=candidates.size()){
            if(target==0) ans.push_back(ds);
            return;
        }
        if(target<0 )return;
        ds.push_back(candidates[i]);
        func(i,ds,ans, candidates,target-candidates[i]);
        ds.pop_back();
        func(i+1,ds,ans, candidates, target);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int> > ans;
        vector<int> ds;
        func(0,ds,ans, candidates, target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    
    public void func(int i, int[] candidates, int target,List<Integer> temp,List<List<Integer>> ans){
        if(i>=candidates.length){
            if(target==0){
                ans.add(new ArrayList<>(temp));
            }
            return;
        } 
        if(target < 0) return;
        temp.add(candidates[i]);
        func(i, candidates, target -candidates[i], temp, ans);
        temp.remove(temp.size()-1);
        func(i+1, candidates, target , temp, ans);
    }
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans =new ArrayList<>();
        List<Integer> temp= new ArrayList<>();
        func(0,candidates, target, temp , ans);
        return ans;

    }
}